pkgs/npm/src/cli.ts

```diff
@@ -15,21 +212,89 @@ function getPathToExecutable(): string {
   try {
     return require.resolve(executablePath);
   } catch (e) {
-    throw new Error(`Could not find application binary at ${executablePath}.`);
+    return null;
   }
 }
 
+function extractCLIVersions(versionInfo: string): {
+  defangCLI: string;
+  latestCLI: string;
+} {
+  // parse the CLI version info
+  // e.g.
+  // Defang CLI:    v0.5.24
+  // Latest CLI:    v0.5.24
+  // Defang Fabric: v0.5.0-643-abcdef012
+  //
+
+  const versionRegex = /\d+\.\d+\.\d+/g;
+  const matches = versionInfo.match(versionRegex);
+
+  if (matches != null && matches.length >= 2) {
+    return {
+      defangCLI: matches[0],
+      latestCLI: matches[1],
+    };
+  } else {
+    throw new Error("Could not extract CLI versions from the output.");
+  }
+}
+
+type VersionInfo = {
+  current: string | null;
+  latest: string | null;
+};
+
+async function getVersionInfo(): Promise<VersionInfo> {
+  let result: VersionInfo = { current: null, latest: null };
+  try {
+    const execPath = getPathToExecutable();
+
+    if (!execPath) {
+      // there is no executable, so we can't get the version info from the CLI
+      const latestVersion = await getLatestVersion();
+
+      return { current: null, latest: latestVersion };
+    }
+
+    // Exec output contains both stderr and stdout outputs
+    const versionInfo = await exec(execPath + " version");
+
+    const verInfo = extractCLIVersions(versionInfo.stdout);
+    result.current = verInfo.defangCLI;
+    result.latest = verInfo.latestCLI;
+  } catch (error) {
+    console.error(error);
+  }
+
+  return result;
+}
+
 // js wrapper to use by npx or npm exec, this will call the defang binary with
 // the arguments passed to the npx line. NPM installer will create a symlink
 // in the user PATH to the cli.js to execute. The symlink will name the same as
 // the package name i.e. defang.
-function run(): void {
+async function run(): Promise<void> {
   try {
+    const { current, latest }: VersionInfo = await getVersionInfo();
+
+    // get the latest version of defang cli if not already installed
+    if (latest != null && (current == null || current != latest)) {
```

Comment:
This is super cool. I love it.

My one question about it is if there is a breaking change that the user isn't expecting. If I run `npx defang@x.y.z` if I understand correctly, starting from this release we will always get the latest release, even if you specify a version number.

Maybe there's a way to check if the user's command included a version number, and if so, use that?

I'm not sure how useful/relevant this is. But it's top of mind for me because I'm supporting someone with an old project these days and having to run and old version of the Firebase CLI.

Reply:
@raphaeltm Good point. Specifying a version should prevent auto upgrading. Let me implement that option.

Reply:
@raphaeltm 
Added a "--use-latest" flag that he cli.js will intercept.
e.g. npx defang --use-latest=false version" to prevent auto-updating.

Only issue is that it's not documented anywhere.

Reply:
Cool! If someone does `npx defang@0.5.1` without the flag, would it actually use `0.5.1`?

Reply:
pre-0.5.25 it would. post-0.5.25 they would still need the flag so it would look like:

**npx defang@0.5.25 --use-latest=false services**

So from now on the user must explicitly say they don't want it to update.
There wasn't a way to get the @version to have the cli.js handle the situation as that's a npx/npm argument that's not passed to the cli wrapper.